Add Private Links support

nprokopic opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3400

**Special notes for your reviewer**:
<!-- Refer to https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/docs/book/src/developers/releasing.md#release-support for more information about which changes are eligible for backport -->

- [ ] cherry-pick candidate

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add Private Links support
```

nprokopic commented:
Hi folks, here is a draft pull request that is adding Private Links support that is described in this issue https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3400.

The feature is intended for clusters with private networking, it adds option to deploy private link for internal API server load balancer, which enables users to access private clusters from anywhere in Azure simply by creating private endpoints to it (CAPZ got support for private endpoints [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3044)). More details in the issue linked above.

What's missing here are tests for the feature, I will add those (in 2 weeks, after vacation), but I wanted to push this draft in case somebody wants to take a look and leave feedback.

We have been testing this at Giant Swarm for a few months now for our private clusters setup and it works quite nicely.

Jont828 commented on azure/services/privatelinks/spec.go:
```diff
@@ -0,0 +1,239 @@
+	for _, frontendIPConfigName := range s.LBFrontendIPConfigNames {
+		frontendIPConfig := network.FrontendIPConfiguration{
+			ID: pointer.String(
+				fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/loadBalancers/%s/frontendIPConfigurations/%s",
```
You should be able to use `FrontendIPConfigID()` in [this file](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/b07d34e72a9ff577c77342b7eb5d6c610c71335c/azure/defaults.go#L246-L249) instead.

Jont828 commented on azure/services/privatelinks/spec.go:
```diff
@@ -0,0 +1,239 @@
+		}
+	}
+
+	if existing != nil {
```
I see why you have this towards the end, but can we move the check to the beginning of the function? That way, it's consistent with the others, and if we get the wrong type we can fail early. In order to check if it's up to date, consider something like this:
```go
if existing != nil {
	existingPrivateLink, ok := existing.(network.PrivateLinkService)
	if !ok {
		return nil, errors.Errorf("%T is not a network.PrivateLinkService", existing)
	}

	privateLinkToCreate, err := s.Parameters(ctx, nil)
	if err != nil {
		return nil, err
	}
	
	if isExistingUpToDate(existingPrivateLink, privateLinkToCreate) {
		return nil, nil
	}
}
```

nprokopic replied:
Good point, will align it.

nprokopic replied:
Update this part of code so there is am early check of `existing` type and now it's similar to your code snipped above (with the remaining required parts of the code there).

Jont828 commented on azure/services/privatelinks/spec.go:
```diff
@@ -0,0 +1,239 @@
+	// NAT IP configuration is not checked as it cannot be changed.
+
+	// Check load balancer configurations
+	var wantedFrontendIDs []string
```
FYI, I believe the PR is blocked on the code coverage tests passing

nprokopic replied:
Code coverage tests should be fine now https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3747#issuecomment-1697387584.

Jont828 commented on azure/services/privatelinks/spec.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+func equalStringSlicesIgnoreOrder(a, b []string) bool {
+	slicesDiff := cmp.Diff(a, b, cmpopts.SortSlices(func(s1, s2 string) bool {
```
WDYT about creating a set for each slice and doing the diff on that instead? That way, it would be in linear time.

Jont828 commented on azure/services/privatelinks/spec.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+func onlyOnePointerIsNil[T any](a, b *T) bool {
+	return a == nil && b != nil || a != nil || b == nil
```
That's a nifty function, I wonder if it's worth adding to a common file as a util.

nprokopic replied:
Yeah, I'll see to move some of these to `util` package.

Jont828 reviewed (commented):
Nice work so far! Besides adding tests, here are some comments

nprokopic commented:
Thanks for the review here @Jont828! :)

I got back from vacation last week, so I'm continuing with this this week, adding tests and addressing the review comments.

Jont828 commented:
@nprokopic Sounds good! Let me know if you need any help.

nprokopic commented:
I'm adding missing tests first, addressed some of the comments along the way. Then after adding all (or most of) required tests I'll address other comments (for the sake of more easily catching potential new errors caused by changes in the implementation).

nprokopic commented on azure/services/privatelinks/mock_privatelinks/doc.go:
```diff
@@ -0,0 +1,23 @@
+
+// Run go generate to regenerate this mock.
+//
+//go:generate ../../../../hack/tools/bin/mockgen -destination client_mock.go -package mock_privatelinks -source ../client.go Client
```
I noticed that mock was generated in other packages even though it was not used. Agreed that it should not be generated, will remove it.

nprokopic commented:
@Jont828 as most of the review comments has been addressed and large amount of tests has been added, I made the PR ready for review now, so it's not in draft state anymore.

I'll address also other remaining comments regarding various optimizations in code.

nprokopic commented:
There is a broken links error:
```
ERROR: 1 dead links found!
[✖] https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-azure#capz-pr-e2e-main → Status: 0
```

Not sure why, as I can successfully open that link.

nprokopic commented on azure/scope/cluster.go:
```diff
@@ -395,6 +396,14 @@ func (s *ClusterScope) SubnetSpecs() []azure.ResourceSpecGetter {
+		// Check if subnet is used for the private link NAT IP
+		for _, privateLink := range s.AzureCluster.Spec.NetworkSpec.APIServerLB.PrivateLinks {
+			for _, ipConfig := range privateLink.NATIPConfigurations {
+				if ipConfig.Subnet == subnetSpec.Name {
```
Performance wise: the two outer for loops are limited, as there can be max 8 private links, and 8 NAT configs per private link. In majority of the cases it will probably be just 1 private link and 1 NAT IP config, unless there is a requirement for some very heavy traffic to the API server.

That being said, optimizing this a bit with maps would also make the code cleaner and more readable, so it could be worth it, so I'll update it.

nprokopic commented on api/v1beta1/azurecluster_validation.go:
```diff
@@ -671,6 +676,131 @@ func validatePrivateEndpoints(privateEndpointSpecs []PrivateEndpointSpec, subnet
+			lbFrontendIPNamesCount := map[string]int{}
+			for j, lbFrontendIPName := range pl.LBFrontendIPConfigNames {
+				lbFrontendIPNameValid := false
+				for _, apiLbFrontendIP := range apiServerLBSpec.FrontendIPs {
```
Refactored so that frontend IPs are stored in a set (`map[string]struct{}{}`). Thanks for the suggestion! :)

nprokopic commented on api/v1beta1/azurecluster_validation.go:
```diff
@@ -671,6 +676,131 @@ func validatePrivateEndpoints(privateEndpointSpecs []PrivateEndpointSpec, subnet
+			}
+			for j, natIPConfig := range pl.NATIPConfigurations {
+				usesValidSubnet := false
+				for _, subnet := range subnets {
```
Refactored so that subnet names are stored in a set (`map[string]struct{}{}`), like for frontend IPs above.

nprokopic commented:
> There is a broken links error:
> 
> ```
> ERROR: 1 dead links found!
> [✖] https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-azure#capz-pr-e2e-main → Status: 0
> ```
> 
> Not sure why, as I can successfully open that link.

`Lint Documentation / Broken Links (pull_request)` check is now green. Looks like a flaky test.

CecileRobertMichon commented:
/milestone v1.11

Jont828 commented:
@nprokopic I just got back from PTO, thanks for working on this. I'll give it a pass today!

Jont828 commented on azure/services/privatelinks/spec.go:
```diff
@@ -0,0 +1,275 @@
+}
+
+// Parameters returns the parameters for the private link.
+func (s *PrivateLinkSpec) parameters() (params network.PrivateLinkService, err error) {
```
Maybe rename to something like `constructParameters()`

Jont828 commented on templates/cluster-template-private-link.yaml:
```diff
@@ -0,0 +1,232 @@
+      - name: ${CLUSTER_NAME}-internal-lb-privatelink
+        lbFrontendIPConfigNames:
+        - ${CLUSTER_NAME}-internal-lb-frontend
+        natIpConfigurations:
```
```suggestion
        natIPConfigurations:
```

Jont828 commented on docs/book/src/topics/api-server-endpoint.md:
```diff
@@ -73,6 +73,45 @@ spec:
+      privateLinks:
+        - lbFrontendIPConfigNames:
+          - lb-private-ip-frontend
+          natIpConfigurations:
```
```suggestion
          natIPConfigurations:
```

Jont828 commented on controllers/azurecluster_reconciler.go:
```diff
@@ -71,6 +72,7 @@ func newAzureClusterService(scope *scope.ClusterScope) (*azureClusterService, er
 			subnets.New(scope),
 			vnetpeerings.New(scope),
 			loadbalancers.New(scope),
+			privatelinks.New(scope),
```
Just to make sure, we have these services in the order of dependency, i.e. VNet must be created before subnet. Is this the right spot for the private link service?

Jont828 reviewed (commented):
Nice work so far! It's tough to drive such a large PR but I think we're getting close to the finish line!

Jont828 commented:
@nprokopic Just a reminder, the v1.11 milestone closes on 9/12 so we might need to move this to the next milestone.

Jont828 commented:
@nprokopic Hey just to follow up, are you still working on this or do you need any help?

CecileRobertMichon commented:
/milestone next

dtzar commented:
/lifecycle stale

nprokopic commented:
> @nprokopic Hey just to follow up, are you still working on this or do you need any help?

Hey folks, sorry for the radio silence, other work priorities got in the way. I'll check out the remaining comments here, and then either me or a another colleague from Giant Swarm will finish the work here.

Will post an update next week.

nawazkh commented:
@nprokopic Thank you for bringing this PR so far. Is there anything we could do to push this beyond the finish line? Let us know over slack or office hours!
